### PR TITLE
Teleporter Deus Vault

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
This adds some hefty power requirements to teleporter related machinery and devices. It also changes the behavior of using the teleporter and will require upgrades or planning to put them to good use.

#### Power Requirements
- Teleporter requires 6 kW of power while active. (previously 0.5 kW)
- Gateway requires 5 kW of power while active. (unchanged)
- Restaurant portal requires 2.5 kW of power while active. (previously 1 kW)

#### Power Use
- Teleporter costs 50 kW to teleport. (previously 0 kW)
- Gateway costs 50 kW to teleport. (previously 0 kW)
- Restaurant portal costs 5 kW to teleport a robot (previously 0 kW)

#### Stock parts
- Teleporter error of margin for target destination radius is affected by the stock parts inside the teleport hub
- The radius when using the teleporter machine in the room is (T1 9x9, T2 6x6, T3 3x3, T4 exact) (previously exact)
- The radius when using the hand tele is +3x3 of whatever the teleporter machine range is (T1 12x12, T2 9x9, T3 6x6, T4 3x3) (previously exact)
- Calibrating a teleporter machine now takes longer (T1 30 sec, T2 20 sec, T3 10 sec, T4 instant) (previously 10 sec)

#### Teleporter Requirements
- The computer console for the teleporter MUST be calibrated
- After calibration, the teleport station MUST be toggled on (aka when the blue portal appears)
- All the equipment must remain powered

#### Hand Teleport Requirements
- The hand tele needs to have an active and calibrated teleporter to use
- Teleporting to None is still possible and unaffected

## Why It's Good For The Game
Perpetual motion machines are not real... _well at least not anymore!_

Initially I was digging through power code when I came across this. The default roundstart station with a basic SM setups uses 400/570 kW of power. That leaves 170 kW of extra power and I was thinking hmmm... wouldn't it be rewarding to make non-critical elements of the station require a big chunk of extra power? We already do it for the Gateway. But the key thing is it is optional and not required for essential station functions.

Teleporter machinery is a good candidate for this. Working on it also allowed me to add tier part upgrades that actually achieve a decent effect. (faster calibration, closer range to target beacon, etc.)

I don't doubt it will annoy some people who are used to have the teleporter instantly function. (calibration/accuracy previously did nothing and were fluff features) Not to mention that teleporting is now more dangerous since you won't land on the beacon exactly. However people can still use it, just not as effectively or without risk.

Ideally both engineering and science have an incentive to provide more power and better stock part upgrades! Win-win.

## Changelog
:cl:
add: Upgrading teleporter machinery stock parts now affects how close in range it can get to the target beacon. (T1 9x9, T2 6x6, T3 3x3, T4 exact) This affects hand teleporters that select the teleporter as well. (hand tele has worse range with an increased +3x3 radius)
balance: Teleporting machinery (Teleporter room, Gateway, Restaurant Portals) now require a substantial amount of power to remain active. When teleporting, the machinery drains a massive amount of power on use.
balance: Teleport computer calibration now takes 30 seconds but can be lowered by upgraded parts. (-10 sec per tier)
balance: Teleporting now requires for the teleport equipment to be properly calibrated, powered, and toggled on. This affects handheld teleporters as well but there is still no requirement to teleport to None locations.
fix: TELEPORTING PERPETUAL MOTION MACHINES NOT REQUIRING POWER
/:cl:

### CODE REVIEW

**Comment by mc-oofert** [comment]
teleporter nerf

**Comment by zoomachina** [comment]
> Upgrading teleporter machinery stock parts now affects how close in range it can get to the target beacon. (T1 9x9, T2 6x6, T3 3x3, T4 exact)

this buffs teleporter as a getaway tool and nerfs teleporter as means of fast travel

**Comment by Livrah** [comment]
tl dr
Why It's Good For The Game: it will annoy some people
^from the PR body

**Comment by moocowswag** [comment]
Dont teleporters very quickly uncalibrate after being calibrated? pretty massive of a hand tele nerf

**Comment by SimplySmiley123** [comment]
doesn't seem to bother hand tele much, but it will piss off engineers royally as well as anyone using teleporters normally without the hand tele, where the hell is any good in this though?

**Comment by zxaber** [comment]
> Dont teleporters very quickly uncalibrate after being calibrated? pretty massive of a hand tele nerf

As they are currently, there is no timer; teleporters lose calibration when someone teleports through. They're good for one use between calibrations.

At a certain part tier, probably T4, they're perma-calibrated.

**Comment by Dorsisdwarf** [comment]
1) Why would anyone ever use the portal mode on the hand teleporter now that it dumps you off randomly in a 12x12 area? That's almost an entire screen of variance, I think a lot of the existing beacons on maps have a pretty good chance of spacing you with that much variance?

2) Is this fully removing uncalibrated teleports? the PR doesn't mention it but I can't see any way that this could happen since you must calibrate the teleporter to turn it on and you removed (technically not documented? should be mentioned in the PR) the decalibration after every teleport. If so, you should remove the flyperson & mothperson transformations as they are no longer possible to obtain

Actually I think that should be added to the pr body and changelog - teleporter calibration being permanent until the destination is changed is a really important part of this PR making even the slightest modicum of sense and it isnt mentioned.

**Comment by SmArtKar** [comment]
I don't think that this is a good change in the slightest, 12x12 is an absurdly huge range, anything beyond 7x7 risks dumping you into another room anyways. With current power implementation what this will result in is teleporters chunking away at other department's power, the only way to make them be a "reward" would be implementing power priorities, and at this point you should probably just implement Mothblocks's power bar allocation system.

This isn't a fix either, please mark your PRs appropriately.

**Comment by Likteer** [comment]
Teleporters requiring calibration removes the joke fly transform. If this passes, I will be inconsolable.

**Comment by ElGitificador** [comment]
wouldnt mind keeping the hand tele nerf since it s got to a point it s one of the most annoying items of the game, but the rest of changes are ridiculously unjustified and the why is it good for the game seems like a joke.

**Comment by Admiral-Gigaton** [comment]
hi, yes, alongside the whole "fuck teleporting into space because the beacon was within 12 tiles of a space tile" thing, 30 seconds to calibrate is, in game, while you are conscious of the fact that you need to wait, a LONG TIME, and you can also get absolutely domed in 10 seconds, miss the escape shuttle entirely in that time (PLUS MAYBE TELEPORT INTO HYPERSPACE BECAUSE THE TELEPORTER WAS T1) or just entirely lose interest because "why should I wait 30 seconds when I can walk to my destination faster than that"

**Comment by necromanceranne** [comment]
I think encouraging more to upgrade the teleporter could be...good, but to be entirely frank, nobody actually uses the teleporter on a regular basis except the captain. So actually getting it upgraded is entirely on the shoulders of the only guy who will use it during the length of a round. Most people don't go upgrade it for both a lack of access to the teleporter itself, as well as it not being a useful machine to upgrade until it is time to escape.

An unupgraded teleporter to a shuttle is practically useless, a 9x9 radius will almost certainly dump you off the shuttle. Nothing short of t4 seems like suicide. Is that...good? I don't think so?

The most meaningful impact this might see is trolling the station by way of teleporter power sinks. 

I can maybe understand wanting to curtail teleporters being able to be utilized largely just to allow for the captain to wisk himself away using a handtele, but the actual machine itself is not the major culprit of this equation. It is entirely the hand teleporter It's a stationary piece of machinery you have to actually go use directly, and is locked up in a command only room.

**Comment by private-tristan** [comment]
I need to check but on some maps aren't there atmos tanks within 12x12 of a teleporter beacon? if you go through a teleporter you could end up stuck in a atmos tank and die before escaping?
Icebox: 
image

**Comment by Exester509** [comment]
> If you care about thermodynamics we should start with the part where space somehow magically freezes you.
Convection heat exchangers working in a vacuum too

**Comment by timothymtorres** [comment]
I'm splitting a good portion of this up into smaller PRs.
